### ISSUE TITLE
Ability to influence policy creation to work around SQS policy size limitations

### ISSUE THREAD

**Issue by danielmarbach** [Particular/NServiceBus.AmazonSQS#527]
As of today we are creating a fairly complex policy on the subscription queue that contains a policy entry per topic subscribed. If an endpoint is handling a larger number of topic we are reaching the SQS policy limits as described below

<img src="https://user-images.githubusercontent.com/174258/86596553-ab6ae680-bf9a-11ea-8bd3-72561ce38162.png" width="315px" />

https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-quotas.html#quotas-policies

```
Error received:
Amazon.SQS.AmazonSQSException: Submitted policy is over max allowed size.
```

The workaround is to disable auto subscribe and create a wild card policy. 

`endpointConfiguration.DisableFeature<AutoSubscribe>();` 

As an example let's imagine Subscriber2 subscribes to Messages.MyEvent and Messages.MyOtherEvent, the following policy would be created

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "180523c7ef74469fbe8b166b8178c2cc",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "sqs:SendMessage",
      "Resource": "arn:aws:sqs:eu-central-1:714232833252:subscriber2",
      "Condition": {
        "ArnLike": {
          "aws:SourceArn": "arn:aws:sns:eu-central-1:714232833252:Messages-MyEvent"
        }
      }
    },
    {
      "Sid": "Sid1594040661602",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "SQS:SendMessage",
      "Resource": "arn:aws:sqs:eu-central-1:714232833252:subscriber2",
      "Condition": {
        "ArnLike": {
          "aws:SourceArn": "arn:aws:sns:eu-central-1:714232833252:Messages-MyOtherEvent"
        }
      }
    }
  ]
}
```

this policy can be deduced to

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "180523c7ef74469fbe8b166b8178c2cc",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "sqs:SendMessage",
      "Resource": "arn:aws:sqs:eu-central-1:714232833252:subscriber2",
      "Condition": {
        "ArnLike": {
          "aws:SourceArn": "arn:aws:sns:eu-central-1:714232833252:Messages-*"
        }
      }
    }
  ]
}
```

it would be a much nicer experience if the behavior could be controlled, for example by trying to deduce a wild card policy based on the matching topic prefixes, so for example

- Messages.MyEvent, Messages.MyOtherEvent --> Messages-*
- Sales.OrderSubmitted, Sales.OrderPayed, Shipping.OrderShipped, Shipping.OrderDidNotArriveInTime --> Sales.* and Shipping.*

and/or expose a function delegate that allows to modify the policy if needed as a fallback before we are propagating the policy.

**Comment by aschenta** [comment]
Error received:
Amazon.SQS.AmazonSQSException: Submitted policy is over max allowed size.

Installers do not need to be disabled. The auto subscribe feature needs to be disabled.

`endpointConfiguration.DisableFeature<AutoSubscribe>();`

**Comment by danielmarbach** [comment]
@aschenta you are absolutely right. Auto Subscription populates the policy and not the installers. I've mixed that up. My bad sorry. I've updated the issue with your addition

**Comment by biddlem** [comment]
we are running into this issue as well, would be great to be able to create a wild card policy; our particular wrinkle with the current workaround is that our endpoint is connected to an NServiceBus Router, and so when we disable the autosubscribe the Router never gets the subscribe requests

**Comment by danielmarbach** [comment]
@biddlem @aschenta @vgmello The SDK recently introduced [breaking changes in regards to policy creation but doesn't offer a replacement yet](https://github.com/aws/aws-sdk-net/issues/1569). Which at this stage would make it very hard for us to allow any kind of sophisticated policy creation overrides for the users. Would it be enough for you if we had something like

```
var transport = transport.UseTransport<>();
var policies = transport.Policies();
policies.UseWildCardPolicy();
```

and then this would create a policy entry per subscriber like

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "sqs:SendMessage",
      "Resource": "arn:aws:sqs:eu-central-1:714232833252:subscriber2",
      "Condition": {
        "ArnLike": {
          "aws:SourceArn": "arn:aws:sns:eu-central-1:714232833252:*"
        }
      }
    },
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "sqs:SendMessage",
      "Resource": "arn:aws:sqs:eu-central-1:714232833252:subscriber1",
      "Condition": {
        "ArnLike": {
          "aws:SourceArn": "arn:aws:sns:eu-central-1:714232833252:*"
        }
      }
    }
  ]
}
```

or would you expect more sophistication like

- Messages.MyEvent, Messages.MyOtherEvent --> Messages-*
- Sales.OrderSubmitted, Sales.OrderPayed, Shipping.OrderShipped, Shipping.OrderDidNotArriveInTime --> Sales.* and Shipping.*

if we want to support both we would need to have some sort of a clear differentiator in name like

```
var transport = transport.UseTransport<>();
var policies = transport.Policies();
policies.UseAccountLevelWildCardPolicy();
```

vs

```
var transport = transport.UseTransport<>();
var policies = transport.Policies();
policies.UseCommonNamespaceWildCardPolicy();
```

or maybe you have better suggestions for the names?

**Comment by biddlem** [comment]
The account level wildcard would certainly work for our situation; ideally the namespace wild card policy would be our preference (more closely aligns to the practical intent of our subscribers)

**Comment by vgmello** [comment]
Our preference would be a generic wildcard policy with the ability to use a custom SNS prefix. (Something like `arn:aws:sns:eu-central-1:714232833252:nsb-*`), this would avoid extracting namespaces from events and also align with the current SNS settings since it currently allows a TopicNamePrefix to be set.
To support advanced scenarios you could consider disabling the inline queue policy creation (if the installers are not on) while keeping the auto subscription. If the installers are off the queues can be created using CF templates outside the NSB process and a policy could be attached to them using the template. I understand that using the installers "flag" might break existing functionally, therefore a new flag like _DisableQueueSubscriptionPolicyCreation_ could be an option.

**Comment by danielmarbach** [comment]
Thanks for all your input. That definitely helps us shaping the next iteration on SQS. It seems there is also good news from the SDK perspective about the policy types, see https://github.com/aws/aws-sdk-net/issues/1732#issuecomment-718921557

We are going to pick this up in the next maintenance release. Can't tell you yet when this will happen but I'll keep you posted.

**Comment by danielmarbach** [comment]
@vgmello @biddlem @aschenta 

Heads up we (@yvesgoeleven @lailabougria and I) are starting to work on making policy creation better and adding the ability to influence the policy creation. We are thinking to allow you to specify multiple values that will be added to the condition. Multiple values on the same condition are evaluated as OR statements and there is no hard limit other than the policy size in how many OR values can be added. These values could mean multiple things such as:

- Account level wildcard (i.ex. `arn:aws:sns:us-east-1:714232877752:*`)
- TopicNamePrefixWildcard (i.ex. `DEV-*`)
- Namespace condition (i.ex. `Sales-*`, `Shipping-*`)

This would result in a policy document like the following

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "4eeb64b1326f4b658dd8b54a3b183491",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "sqs:SendMessage",
      "Resource": "arn:aws:sqs:us-east-1:714232833252:AT01SHG3XEJ0EIIYBIRIANUWBaseEventFrom2Publishers-Subscriber1",
      "Condition": {
        "ArnLike": {
          "aws:SourceArn": [
              "arn:aws:sns:us-east-1:714232877752:*",
              "arn:aws:sns:us-east-1:714232877752:DEV-*", 
              "arn:aws:sns:us-east-1:714232877752:Sales-*",
              "arn:aws:sns:us-east-1:714232877752:Shipping-*"
          ]
        }
      }
    }
  ]
}
```

an API could look like the following:

```csharp
var policy = transport.Policies();
policy.AddAccountCondition();
policy.AddTopicNamePrefixCondition(); // extracted from transport.TopicNamePrefix("DEV-")
policy.AddNamespaceCondition("Sales."); // dots turned to dashes and if prefix set it would be taken into account
policy.AddNamespaceCondition("Shipping."); // dots turned to dashes and if prefix set it would be taken into account
```

we think this scales better than the current approach and provides more flexibility. Based on the new policy design we are thinking about actually replacing the default exact policy from

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "180523c7ef74469fbe8b166b8178c2cc",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "sqs:SendMessage",
      "Resource": "arn:aws:sqs:eu-central-1:714232877752:subscriber2",
      "Condition": {
        "ArnLike": {
          "aws:SourceArn": "arn:aws:sns:eu-central-1:714232877752:Messages-MyEvent"
        }
      }
    },
    {
      "Sid": "Sid1594040661602",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "SQS:SendMessage",
      "Resource": "arn:aws:sqs:eu-central-1:714232877752:subscriber2",
      "Condition": {
        "ArnLike": {
          "aws:SourceArn": "arn:aws:sns:eu-central-1:714232877752:Messages-MyOtherEvent"
        }
      }
    }
  ]
}
```

to

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "4eeb64b1326f4b658dd8b54a3b183491",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "sqs:SendMessage",
      "Resource": "arn:aws:sqs:us-east-1:714232877752:AT01SHG3XEJ0EIIYBIRIANUWBaseEventFrom2Publishers-Subscriber1",
      "Condition": {
        "ArnLike": {
          "aws:SourceArn": [
              "arn:aws:sns:eu-central-1:714232877752:Messages-MyEvent",
              "arn:aws:sns:eu-central-1:714232877752:Messages-MyOtherEvent"
          ]
        }
      }
    }
  ]
}
```

this would allow us to scale up to 5 times more events per queue and of course even more when using the advanced wildcard capability. We are thinking about replacing the first matching statement (matching = the one that looks like ours ;) ) with that new OR value policy and leave the other statements intact just in case any custom ones have been added. By doing so we can also make sure there is no message loss scenario while updating the policy. As an additional help we could log WARN statements that we think could be removed from the policy. The downside of this approach is that you would be in charge of cleanup the policies. 

What do you think? Can you shoot holes into this approach? 

In addition, we couldn't find any documentation that would indicate there are performance issues when evaluating OR values in comparison to statement evaluation, if you have anything to share with us that would be appreciated.

**Comment by danielmarbach** [comment]
This has been fixed and adjusted in 5.3.0. The announcement is available at

https://discuss.particular.net/t/nservicebus-amazonsqs-5-3-0-minor-release-available/2261

the documentation for the policy configuration options at 

https://docs.particular.net/transports/sqs/configuration-options?version=sqs_5.3#policy
